File: performance_tests/real_sqr_fold.cpp
#include <iostream>
#include <numeric>
#include <vector>

using Real = double;

Real foo(Real x, const std::vector<Real>& s) {
    return std::accumulate(
        s.begin(), s.end(), x,
        [](const Real sum, const Real si) { return sum + si * si; });
}

int main() {
    std::vector<Real> s{1.5, 2.5, 3.5, 4.5, 5.5, 1.5, 2.5, 3.5, 4.5, 5.5,
                        1.5, 2.5, 3.5, 4.5, 5.5, 1.5, 2.5, 3.5, 4.5, 5.5};

    Real x = Real(0);

    std::cin >> x;

    for (size_t i = 0; i < 100'000'000; ++i) {
        x = foo(x, s);
    }

    std::cout << std::fixed << x << std::endl;

    return 0;
}

File: performance_tests/real_sqr_fold.rs
use std::io;

type Real = f64;

pub fn foo(x: Real, s: &Vec::<Real>) -> Real {
    s.iter().fold(x, |sum, si| sum + si * si)
}

fn main() {
    let s: Vec::<Real> = vec![ 1.5, 2.5, 3.5, 4.5, 5.5, 1.5, 2.5, 3.5, 4.5, 5.5, 1.5, 2.5, 3.5, 4.5, 5.5, 1.5, 2.5, 3.5, 4.5, 5.5 ];

    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();

    let mut x: Real = line.trim().parse().unwrap();

    for _i in 0..100_000_000 {
        x = foo(x, &s);
    }
    
    println!("{}", x)
}
